src/Symfony/Component/Translation/Bridge/Phrase/PhraseProvider.php

```diff
@@ -0,0 +1,276 @@
+                if (false === $this->readConfig->isFallbackLocaleEnabled()) {
+                    $headers = $response->getHeaders(false);
+                    $item->set(new PhraseCachedResponse($headers['etag'][0], $headers['last-modified'][0], $content));
+                    $this->cache->save($item);
+                }
+            }
+        }
+
+        $event = new PhraseReadEvent($translatorBag);
+        $this->dispatcher->dispatch($event);
+
+        return $event->getBag();
+    }
+
+    public function write(TranslatorBagInterface $translatorBag): void
+    {
+        \assert($translatorBag instanceof TranslatorBag);
+
+        $event = new PhraseWriteEvent($translatorBag);
+        $this->dispatcher->dispatch($event);
+
+        $catalogues = $event->getBag()->getCatalogues();
+
+        foreach ($catalogues as $catalogue) {
+            foreach ($catalogue->getDomains() as $domain) {
+                if (0 === \count($catalogue->all($domain))) {
+                    continue;
+                }
+
+                $phraseLocale = $this->getLocale($catalogue->getLocale());
+
+                $content = $this->xliffFileDumper->formatCatalogue($catalogue, $domain, ['default_locale' => $this->defaultLocale]);
+                $filename = sprintf('%d-%s-%s.xlf', date('YmdHis'), $domain, $catalogue->getLocale());
+
+                $fields = array_merge($this->writeConfig->setTag($domain)->setLocale($phraseLocale)->getOptions(), ['file' => new DataPart($content, $filename, 'application/xml')]);
+
+                $formData = new FormDataPart($fields);
+
+                $response = $this->httpClient->request('POST', 'uploads', [
+                    'body' => $formData->bodyToIterable(),
+                    'headers' => $formData->getPreparedHeaders()->toArray(),
+                ]);
+
+                if (201 !== $statusCode = $response->getStatusCode()) {
+                    $this->logger->error(sprintf('Unable to upload translations for domain "%s" to phrase: "%s".', $domain, $response->getContent(false)));
+
+                    $this->throwProviderException($statusCode, $response, 'Unable to upload translations to phrase.');
+                }
+            }
+        }
+    }
+
+    public function delete(TranslatorBagInterface $translatorBag): void
+    {
+        $keys = [[]];
+
+        foreach ($translatorBag->getCatalogues() as $catalogue) {
+            foreach ($catalogue->getDomains() as $domain) {
+                $keys[] = array_keys($catalogue->all($domain));
+            }
+        }
+
+        $keys = array_unique(array_merge(...$keys));
+        $names = array_map(static fn ($v): ?string => preg_replace('/([\s:,])/', '\\\\\\\\$1', $v), $keys);
+
+        foreach ($names as $name) {
+            $response = $this->httpClient->request('DELETE', 'keys', [
+                'query' => [
+                    'q' => 'name:'.$name,
```

Comment:
did you try concatenating many keys at once?
that's a strange API - using the query string, etc...

Reply:
btw, don't you also need to select by domain (tag)?

Reply:
as for tags: they are nothing but a label in Phrase while a key is unique within a project. so when you delete a key, it will be deleted regardless of the tags attached to it

the only way to batch delete keys is to have the Phrase identifier related to that key. these can be retieved at their [get keys](https://developers.phrase.com/api/#get-/projects/-project_id-/keys) endpoint, but that response is paginated with a max of 100 objects per response which would potentially result in a large amount of requests when working on a project with thousands of translation keys.

so the option chosen is to delete keys by name, which has the downside that it requires one request per key to be deleted as it doesn't support multiple keys at once.

Reply:
hmmm, thinking of it: an alternative might be to first do a file upload with the keys to be deleted and tag them with an unique string after which a delete can be done using that tag as identifier for keys to be deleted.

not sure whether that's a safe way to go tho; i can imagine a scenario where the file upload is still being processed by phrase when you fire a request to delete keys based on a value used in that upload. might be a slippery slope...

Reply:
> but that response is paginated with a max of 100 objects per response which would potentially result in a large amount of requests when working on a project with thousands of translation keys.

I guess it's not a big deal. I mean, it's ok to send lots of requests if needed, we already do it in other providers. The point is to carefully handle the response to be sure to continue the process.

Moreover, I think it's very rare in a project (even more on a large project) to delete a big amount of translations at once. And there is always a workaround for users: delete small batches of keys, one by one (I thinks it's ok to mention this as warning in the documentation)

Reply:
the thing with phrase is though that their api is rate limited so lots of requests will potentially consume a big chunk of that limit while most of the responses are probably left unused.

agreed on the workaround bit and adding a warning to the documentation

Reply:
Let's do with a deletion by name for now. If someone need to improve this part to delete more keys at a time, we will look into it again